opendevin/controller/agent_controller.py

```diff
@@ -150,11 +155,42 @@ def get_task_state(self):
     async def notify_task_state_changed(self):
         await self._run_callbacks(TaskStateChangedAction(self._task_state))
 
+    async def add_user_message(self, message: UserMessageObservation):
+        if self.state is None:
+            return
+
+        if self._task_state == TaskState.AWAITING_USER_INPUT:
+            self._await_user_message_queue.put_nowait(message)
+
+            # set the task state to running
+            self._task_state = TaskState.RUNNING
+            await self.notify_task_state_changed()
+
+        elif self._task_state == TaskState.RUNNING:
+            self.add_history(NullAction(), message)
+
+        else:
+            raise ValueError(f'Task (state: {self._task_state}) is not in a state to add user message')
+
+    async def wait_for_user_input(self) -> UserMessageObservation:
+        self._task_state = TaskState.AWAITING_USER_INPUT
+        await self.notify_task_state_changed()
+        # wait for the next user message
+        if len(self.callbacks) == 0:
+            logger.info('Use STDIN to request user message as no callbacks are registered', extra={'msg_type': 'INFO'})
+            message = input('Request user input >> ')
+            user_message_observation = UserMessageObservation(message)
+        else:
+            user_message_observation = await self._await_user_message_queue.get()
+            self._await_user_message_queue.task_done()
+        return user_message_observation
+
```

Comment:
@rbren @enyst here's the implemtation of "wait for user info" logic i discussed in https://opendevin.slack.com/archives/C06QKSD9UBA/p1713775908496249

Reply:
When the server side waiting for a user input, but user just close the window, how we process with this? What if it's just a  network break for a while, and the browser side come back again seconds later?  @xingyaoww

Reply:
Good questions!

We probably need some logic to clean up sessions that have been idle for > x seconds